Add blogpost for z score normalization technique (#3720)
opensearch-project/project-website: _posts/2025-03-31-zscore-hybrid-search.md

            }
        }
    ]
}
```

Next, create a hybrid query and apply the pipeline to the query:

```json
POST my_index/_search?search_pipeline=z_score-pipeline
{
   "query": {
     "hybrid": [
         {}, // First Query
         {}, // Second Query
         ... // Other Queries
     ] 
   }
}
```

## Benchmarking z-score normalization performance

Benchmark experiments were conducted on an OpenSearch cluster with one **r6g.8xlarge** instance as the coordinator node and three **r6g.8xlarge** instances as data nodes. To evaluate z-score normalization comprehensively, we measured two key metrics across five distinct datasets.  

For more information about the datasets used, see [Datasets](https://public.ukp.informatik.tu-darmstadt.de/thakur/BEIR/datasets/).

### Sample queries and passages

The following table provides sample queries and passages for each dataset.

|Dataset	|Sample query	|Sample passage	|
|:---	|:---	|:---	|
|SciDocs	|CFD Analysis of Convective Heat Transfer Coefficient on External Surfaces of Buildings	|`This paper provides an overview of the application of CFD in building performance simulation for the outdoor environment, focused on four topics...`	|

> |FiQA	|“Business day” and “due date” for bills	|`I don't believe Saturday is a business day either. When I deposit a check at a bank's drive-in after 4pm Friday, the receipt tells me it will credit as if I deposited on Monday. If a business' computer doesn't adjust their billing to have a weekday due date ...	`|

Reviewer: ```suggestion
|FiQA	|"Business day" and "due date" for bills	|`I don't believe Saturday is a business day either. When I deposit a check at a bank's drive-in after 4pm Friday, the receipt tells me it will credit as if I deposited on Monday. If a business' computer doesn't adjust their billing to have a weekday due date ...	`|
```

|touche2020	|Is a college education worth it?	|`The resolution used by Pro *assumes* that Australia isn't already a 'significant' country - however, in actual reality, it is. Firstly we should clarify what significance means: 1.a the state or quality of being significant1.b of consequence or..`	|

Search relevance was quantified using the industry-standard normalized discounted cumulative gain at rank 10 (NDCG@10). We also tracked system performance using search latency measurements. This setup provided a strong foundation for evaluating both search quality and operational efficiency.

### NDCG@10

The following table compares NDCG@10 scores for hybrid search using min-max normalization and z-score normalization, showing the percent difference in search relevance across five datasets.

|Dataset	|Hybrid (min-max)	|Hybrid (z-score)	|Percent difference	|
|---	|---	|---	|---	|
|SciDocs	|0.1591	|0.1633	|+2.45%	|
|FiQA	|0.2747	|0.2768	|+0.77%	|
|NQ	|0.3665	|0.374	|+2.05%	|
|ArguAna	|0.4507	|0.467	|+3.62%	|
|touche2020	|0.841	|0.8542	|+1.54%	|
|	|	|Average	|2.08%	|

### Search latency

> The following table presents search latency measurements in milliseconds at different percentiles (p50, p90, and p99) for hybrid search using min-max and z-score normalization. The *Percent difference* columns indicate the relative performance impact between these methods.

Reviewer: ```suggestion
The following table presents search latency measurements in milliseconds at different percentiles (p50, p90, and p99) for hybrid search using min-max and z-score normalization. The *Percent difference* columns indicate the relative performance impact of these methods.
```

---
layout: post
title:  Z Score Normalization Technique for Hybrid Search
authors:
  - kazabdu
  - gaievski
date: 2025-03-31
has_science_table: true
categories:
  - technical-posts
meta_keywords: z score normalization, OpenSearch 3.0-beta, neural search plugin, hybrid search, relevance ranking, search normalization, k-nn search, L2 normalization, how reciprocal rank fusion works
meta_description: Learn about z score normalization using the Neural Search plugin in OpenSearch 3.0-beta. Discover how this new approach to hybrid search merges results from multiple query sources for improved relevance.
---

In the world of search engines and machine learning, data normalization plays a crucial role in ensuring fair and accurate comparisons between different features or scores. 

> Hybrid query uses multiple normalization techniques for preparing final results, main two types are score based normalization and rank base combination. In score base normalization, min-max normalization doesn’t work well with outliers (Outliers are data points that significantly differ from other observations in a dataset. 

Reviewer 2: Can we mention that l2 is also an option for score based normalization and maybe a weakness of that one?

Author: l2 is not affected by outliers by that much as it covers relative magnitudes of features. And this blogpost is moreover comparison between with min max

In the context of normalization techniques like Min-Max scaling and Z-score (Standard Score) normalization, outliers can have a substantial impact on the results). In this blogpost we would introduce another normalization technique called as z-score which was added in OpenSearch 3.0.0-beta release. 
Let's dive into what Z-score normalization is, why it's important, and how it's being used in OpenSearch.

## What is Z-Score Normalization?

Z-score normalization, also known as standardization, is a method of scaling data using mean and standard deviation. The formula for calculating the Z-score is:
Z = (X - μ) / σ
Where:

* X is the original value
* μ is the mean of the population
* σ is the standard deviation of the population

## When to use Z Score?

> Considering your index’s structure can help you decide which one to choose since each has advantages of its own. If your documents are more similar to one another and the top-k results of a typical query return documents that are very similar to one another and clustered together within the index, as seen in the graph below, Min-Max may be a better option.

Reviewer 2: Same as above, is there any way to compare to l2 instead of purely comparing against min-max?

Author: Replied above

![Image for min max distribution](/assets/media/blog-images/2025-03-31-zscore-hybrid-search/blogpost1.jpg){: .img-fluid}

However, Z-Score is more suited if the results are more evenly distributed and have some characteristics of a normal distribution, as shown in the example below.

![Image for zscore distribution](/assets/media/blog-images/2025-03-31-zscore-hybrid-search/blogpost2.jpg){: .img-fluid}

The basic flow to use between min max and z score looks like below:

![Image for flow](/assets/media/blog-images/2025-03-31-zscore-hybrid-search/blogpost3.png){: .img-fluid}

### How to use Z Score?

To use z_score, create a search pipeline and specify `z_score` as the `technique`:

```
PUT /_search/pipeline/z_score-pipeline
{
    "description": "Zscore processor for hybrid search",
    "phase_results_processors": [
        {
            "normalization-processor": {
                "normalization": {
                    "technique": "z_score"
                },
                "combination": {

>                     "technique": "arithmetic_mean"

Reviewer 3: I would give an example without combination technique, we support only one AM, this may give false impression that we support same subset of three techniques, as for other normalization methods.

                }
            }
        }
    ]
}
```

Next, create a hybrid query and apply the pipeline to it

```

POST my_index/_search?search_pipeline=z_score-pipeline
{
   "query": {
     "hybrid": [
         {}, // First Query
         {}, // Second Query
         ... // Other Queries
     ] 
   }
}
```

## Benchmarking Z Score performance

> Benchmark experiments were conducted using an OpenSearch cluster consisting of a single r6g.8xlarge instance as the coordinator node, along with three r6g.8xlarge instances as data nodes. To assess Z Score’s performance comprehensively, we measured three key metrics across four distinct datasets. For information about the datasets used, see [Datasets](https://public.ukp.informatik.tu-darmstadt.de/thakur/BEIR/datasets/).

Reviewer 3: Can you change the ending of this paragraph, typically for blogs we try to put more info regarding datasets, like for instance query length avg/med, doc length avg/med, number of queries. Not all of that is present in beir repo. 
You can add a table just for those 4 you used for experiments, similar to how it's done in this blog https://opensearch.org/blog/semantic-science-benchmarks, and also keep the reference to beir repo for other details

> Search relevance was quantified using the industry-standard Normalized Discounted Cumulative Gain at rank 10 (NDCG@10). We also tracked system performance using search latency measurements and relevancy. This setup provided a strong foundation for evaluating both search quality and operational efficiency.

Reviewer 2: Nit: Should we just say `using search latency measurements` since relevancy isn't related to tracking system performance?

> Search relevance was quantified using the industry-standard Normalized Discounted Cumulative Gain at rank 10 (NDCG@10). We also tracked system performance using search latency measurements and monitored CPU utilization to analyze resource consumption during the experiments. This setup provided a strong foundation for evaluating both search quality and operational efficiency.

Reviewer 2: Is there supposed to be a table for the CPU utilization?

Author: Since both are score based techniques, cpu utilization won't have major difference

Reviewer 2: Can we either remove the monitored CPU utilization part here then or mention that in the conclusions?

Author: Done

### Conclusions

Our benchmark experiments highlight the following advantages and trade-offs of Z-score normalization compared to min-max normalization in hybrid search approaches:

**Search quality (measured using NDCG@10 across four datasets)**:

* Z-score normalization shows a modest improvement in search quality, with an average increase of 1.765% in NDCG@10 scores.

> * This suggests that Z-score normalization may provide slightly better relevance in search results compared to min-max normalization.

Reviewer 3: mention that min_max is out default, this will give a reason why we're comparing z-score with it

**Trade-offs**:

* There's a slight trade-off between search quality and latency. Z-score normalization offers a small improvement in search relevance (1.765% increase in NDCG@10) at the cost of a marginal increase in latency (0.50% to 0.72% across different percentiles).

**Consistency**:

* The impact of Z-score normalization varies across datasets, with some showing more significant improvements in search quality than others.
* Latency impact is relatively consistent across datasets, with most showing small increases or no change

**Overall assessment**:

* Z-score normalization provides a modest improvement in search quality with a negligible impact on latency.
* The choice between Z-score and min-max normalization may depend on specific use cases, with Z-score potentially being preferred when even small improvements in search relevance are valuable and the slight latency increase is acceptable.

These findings suggest that Z-score normalization could be a viable alternative to min-max normalization in hybrid search approaches, particularly in scenarios where optimizing search relevance is a priority and the system can tolerate minimal latency increases

## What’s next?

> We are also expanding OpenSearch’s hybrid search capabilities beyond z score by planning the following improvements to our normalization framework:

Reviewer 3: maybe `enhancement` is the better word, improvements implies that we're fixing something that is already there. This is not the case for custom functions, this is totally new area

**Search quality (measured using NDCG@10 across four datasets)**:

* Z-score normalization shows a modest improvement in search quality, with an average increase of 2.08% in NDCG@10 scores.
* This suggests that Z-score normalization may provide slightly better relevance in search results compared to the default normalization technique min-max.

**Latency impact**:

* Z-score normalization shows a small increase in latency across different percentiles, as shown in the following table:

|Latency percentile	|Percent difference	|
|---	|---	|
|p50	|0.70%	|
|p90	|0.50%	|
|p99	|0.50%	|

* The positive percentages indicate that Z-score normalization has slightly higher latency compared to min-max normalization, but the differences are minimal (less than 1% on average).

**Trade-offs**:

* There's a slight trade-off between search quality and latency. Z-score normalization offers a improvement in search relevance (2.08% increase in NDCG@10) at the cost of a marginal increase in latency (0.50% to 0.72% across different percentiles).

> **Consistency**:

Reviewer 3: I'm not sure if we need this section, afaik for some datasets they were no improvement in relevancy, we probably can't claim improvement is consistent. If you want to mention increase in latency you probably can do it in the "Overall assessment" section that follows.

Author: Removed the section

* The impact of Z-score normalization varies across datasets, with some showing more significant improvements in search quality than others.
* Latency impact is relatively consistent across datasets, with most showing small increases or no change

**Overall assessment**:

* Z-score normalization provides a modest improvement in search quality with a negligible impact on latency.

> * The choice between Z-score and min-max normalization may depend on specific use cases, with Z-score potentially being preferred when even improvements in search relevance are valuable and the slight latency increase is acceptable.

Reviewer 3: looks like some word is missing after "even"

Author: Good catch. Pushed the change

These findings suggest that Z-score normalization could be a viable alternative to min-max normalization in hybrid search approaches, particularly in scenarios where optimizing search relevance is a priority and the system can tolerate minimal latency increases

## What’s next?

We are also expanding OpenSearch’s hybrid search capabilities beyond z score by planning the following enhancements to our normalization framework:

**Custom normalization functions**: Enables you to define your own normalization logic and allows fine-tuning of search result rankings. For more information, see [this issue](https://github.com/opensearch-project/neural-search/issues/994).

These enhancements will provide more control over search result ranking while ensuring reliable and consistent hybrid search outcomes. Stay tuned for more information!

### References

1. https://www.codecademy.com/article/normalization
2. https://towardsdatascience.com/hybrid-search-2-0-the-pursuit-of-better-search-ce44d6f20c08/

> 3. https://github.com/opensearch-project/neural-search/issues/1209

Reviewer 3: Can you put few words description for each item and make them a link, e.g. "[RFC] Z Score Normalization Technique for Normalization Processor" for the last link

## How to use z-score normalization

To use z-score normalization, create a search pipeline and specify `z_score` as the `technique`:

```json
PUT /_search/pipeline/z_score-pipeline
{
    "description": "Zscore processor for hybrid search",
    "phase_results_processors": [
        {
            "normalization-processor": {
                "normalization": {
                    "technique": "z_score"
                }
            }
        }
    ]
}
```

Next, create a hybrid query and apply the pipeline to the query:

```json
POST my_index/_search?search_pipeline=z_score-pipeline
{
   "query": {
     "hybrid": [
         {}, // First Query
         {}, // Second Query
         ... // Other Queries
     ] 
   }
}
```

## Benchmarking z-score normalization performance

Benchmark experiments were conducted on an OpenSearch cluster with one **r6g.8xlarge** instance as the coordinator node and three **r6g.8xlarge** instances as data nodes. To evaluate z-score normalization comprehensively, we measured two key metrics across five distinct datasets.  

## Benchmarking Z Score for Development Environment

> Benchmark experiments were conducted using an OpenSearch cluster consisting of a single r6g.8xlarge instance as the coordinator node, along with three r6g.8xlarge instances as data nodes with 1 shard. To assess Z Score’s performance comprehensively, we measured two key metrics across five distinct datasets. For more information about the datasets used, see [Datasets](https://public.ukp.informatik.tu-darmstadt.de/thakur/BEIR/datasets/).

Reviewer: ```suggestion
Benchmark experiments were conducted using an OpenSearch cluster consisting of a single r6g.8xlarge instance as the coordinator node, along with three r6g.8xlarge instances as data nodes with 1 shard. To assess z -score normalization performance comprehensively, we measured two key metrics across five distinct datasets. For more information about the datasets used, see [Datasets](https://public.ukp.informatik.tu-darmstadt.de/thakur/BEIR/datasets/).
```

  |Latency percentile	|Percent difference	|
  |---	|---	|
  |p50	|0.70%	|
  |p90	|0.50%	|
  |p99	|0.50%	|

  * The positive percentages indicate that z-score normalization has slightly higher latency compared to min-max normalization, but the differences are minimal (less than 1% on average).

- **Trade-offs**:

  * There's a slight trade-off between search quality and latency. Z-score normalization offers an improvement in search relevance (2.08% increase in NDCG@10) at the cost of a marginal increase in latency (0.50% to 0.72% across different percentiles).

- **Overall assessment**:

  * Z-score normalization provides a modest improvement in search quality with a negligible impact on latency.
  * The choice between z-score and min-max normalization may depend on your use case. If search relevance is a priority, z-score normalization may be the better option, especially if your system can tolerate a small latency increase.

## Benchmarking Z Score for Production Scenario

Benchmark experiments were conducted using an OpenSearch cluster consisting of a single r6g.8xlarge instance as the coordinator node, along with three r6g.8xlarge instances as data nodes with 12 shards.

> ### NDCG@mean

Reviewer: Please add an intro sentence for the table (similar to the one in the section above beginning with "The following table.."

> |dataset	|Hybrid (min max)	|Hybrid (z score)	|Percent diff	|

Reviewer: ```suggestion
|Dataset	|Hybrid (min-max)	|Hybrid (z-score)	|Percent difference	|
```

|---	|---	|---	|---	|

> |fiqa	|0.3105	|0.3120	|+0.48%	|

Reviewer: ```suggestion
|FiQA	|0.3105	|0.3120	|+0.48%	|
```

> |nq	|0.4563	| 0.4540	|-0.50%|

Reviewer: ```suggestion
|NQ	|0.4563	| 0.4540	|-0.50%|
```

> |arguana	|0.4713	|0.4700	|-0.28%	|

Reviewer: ```suggestion
|ArguAna	|0.4713	|0.4700	|-0.28%	|
```

|touche2020	|0.3788	|0.3719	|-1.82%	|
|	|	|Average	|-0.53%	|

> ### P@mean

Reviewer: Please provide an intro sentence for this table.

> |dataset	|Hybrid (min max)	|Hybrid (z score)	|Percent diff	|

Reviewer: ```suggestion
|Dataset	|Hybrid (min-max)	|Hybrid (z-score)	|Percent difference	|
```

|---	|---	|---	|---	|

> |fiqa	|0.0921	|0.0925	|+0.43%	|

Reviewer: ```suggestion
|FiQA	|0.0921	|0.0925	|+0.43%	|
```

> |nq	|0.0943	| 0.0935	|-0.85%|

Reviewer: ```suggestion
|NQ| 0.0943	| 0.0935	|-0.85%|
```

> |arguana	|0.0878	|0.0878	|0%	|

Reviewer: ```suggestion
|ArguAna	|0.0878	|0.0878	|0%	|
```

---
layout: post
title:  Introducing the z-score normalization technique for hybrid search
authors:
  - kazabdu
  - gaievski
  - minalsha
  - kolchfa

> date: 2025-03-22

Reviewer 4: Update publish date to 2025-04-28

has_science_table: true
categories:
  - technical-posts
meta_keywords: z-score normalization, OpenSearch 3.0-beta1, neural search plugin, hybrid search, relevance ranking, search normalization, k-nn search, L2 normalization, how reciprocal rank fusion works

> meta_description: Learn about z-score normalization in OpenSearch 3.0-beta1. Discover how this new approach to hybrid search merges results from multiple query sources for improved relevance.

Reviewer 4: Please update the mta wit the following: 

_meta_keywords:_ z-score normalization, hybrid search, data normalization, OpenSearch search latency, search relevance, hybrid search score normalization

_meta_description:_ Explore z-score normalization in OpenSearch 3.0-beta1 for hybrid search. Learn how this technique compares to min-max normalization through benchmarks on search relevance and performance.

Author: @pajuric thanks for the review. Pushed the change

|FiQA	|"Business day" and "due date" for bills	|`I don't believe Saturday is a business day either. When I deposit a check at a bank's drive-in after 4pm Friday, the receipt tells me it will credit as if I deposited on Monday. If a business' computer doesn't adjust their billing to have a weekday due date ...	`|
|NQ |what is non controlling interest on balance sheet |`In accounting, minority interest (or non-controlling interest) is the portion of a subsidiary corporation's stock that is not owned by the parent corporation. The magnitude of the minority interest in the subsidiary company is generally less than 50% of outstanding shares, or the corporation would generally cease to be a subsidiary of the parent`|
|ArguAna	|Poaching is becoming more advanced A stronger, militarised approach is needed as poaching is becoming ...	|`Tougher protection of Africa\u2019s nature reserves will only result in more bloodshed. Every time the military upgrade their weaponry, tactics and logistic, the poachers improve their own methods to counter ...`	|

> |touche2020	|Is a college education worth it?	|`The resolution used by Pro *assumes* that Australia isn't already a 'significant' country - however, in actual reality, it is. Firstly we should clarify what significance means: 1.a the state or quality of being significant1.b of consequence or..`	|

Reviewer 5: ```suggestion
|Touche-2020	|Is a college education worth it?	|`The resolution used by Pro *assumes* that Australia isn't already a 'significant' country - however, in actual reality, it is. Firstly we should clarify what significance means: 1.a the state or quality of being significant1.b of consequence or..`	|
```

> |Touche2020	|Is a college education worth it?	|`The resolution used by Pro *assumes* that Australia isn't already a 'significant' country - however, in actual reality, it is. Firstly we should clarify what significance means: 1.a the state or quality of being significant1.b of consequence or..`	|

Reviewer 5: ```suggestion
|Touche-2020	|Is a college education worth it?	|`The resolution used by Pro *assumes* that Australia isn't already a 'significant' country - however, in actual reality, it is. Firstly we should clarify what significance means: 1.a the state or quality of being significant1.b of consequence or..`	|
```

Search relevance was quantified using the industry-standard normalized discounted cumulative gain at rank 10 (NDCG@10). We also tracked system performance using search latency measurements. This setup provided a strong foundation for evaluating both search quality and operational efficiency.

### NDCG@10

The following table compares NDCG@10 scores for hybrid search using min-max normalization and z-score normalization, showing the percent difference in search relevance across five datasets.

|Dataset	|Hybrid (min-max)	|Hybrid (z-score)	|Percent difference	|
|---	|---	|---	|---	|
|SciDocs	|0.1591	|0.1633	|+2.45%	|
|FiQA	|0.2747	|0.2768	|+0.77%	|
|NQ	|0.3665	|0.374	|+2.05%	|
|ArguAna	|0.4507	|0.467	|+3.62%	|

> |Touche2020	|0.841	|0.8542	|+1.54%	|

Reviewer 5: ```suggestion
|Touche-2020	|0.841	|0.8542	|+1.54%	|
```

|	|	|Average	|2.08%	|

### Search latency

The following table presents search latency measurements in milliseconds at different percentiles (p50, p90, and p99) for hybrid search using min-max and z-score normalization. The *Percent difference* columns indicate the relative performance impact of these methods.

> <table> <tr> <th></th> <th colspan="3"><b>p50</b></th> <th colspan="3"><b>p90</b></th> <th colspan="3"><b>p99</b></th> </tr> <tr> <td></td> <td><b>Hybrid (min-max)</b></td> <td><b>Hybrid (z-score)</b></td> <td><b>Percent difference</b></td> <td><b>Hybrid (min-max)</b></td> <td><b>Hybrid (z-score)</b></td> <td><b>Percent difference</b></td> <td><b>Hybrid (min-max)</b></td> <td><b>Hybrid (z-score)</b></td> <td><b>Percent difference</b></td> </tr> <tr> <td>SciDocs</td> <td>76.25</td> <td>77.5</td> <td>1.64%</td> <td>99</td> <td>100.5</td> <td>1.52%</td> <td>129.54</td> <td>133.04</td> <td>2.70%</td> </tr> <tr> <td>FIQA</td> <td>80</td> <td>81</td> <td>1.25%</td> <td>104.5</td> <td>105</td> <td>0.48%</td> <td>123.236</td> <td>124</td> <td>0.62%</td> </tr> <tr> <td>NQ</td> <td>117</td> <td>117</td> <td>0%</td> <td>140</td> <td>140</td> <td>0%</td> <td>166.74</td> <td>165.24</td> <td>-0.90%</td> </tr> <tr> <td>ArguAna</td> <td>349</td> <td>349</td> <td>0%</td> <td>382</td> <td>382</td> <td>0%</td> <td>417.975</td> <td>418.475</td> <td>0.12%</td> </tr> <tr> <td>Touche2020</td> <td>77</td> <td>77.5</td> <td>0.64%</td> <td>100</td> <td>100.5</td> <td>0.50%</td> <td>140</td> <td>140</td> <td>0%</td> </tr> <tr> <td></td> <td></td> <td><b>Average:</b></td> <td>0.70%</td> <td></td> <td><b>Average:</b></td> <td>0.50%</td> <td></td> <td><b>Average:</b></td> <td>0.50%</td> </tr> </table>

Reviewer 5: ```suggestion
<table> <tr> <th></th> <th colspan="3"><b>p50</b></th> <th colspan="3"><b>p90</b></th> <th colspan="3"><b>p99</b></th> </tr> <tr> <td></td> <td><b>Hybrid (min-max)</b></td> <td><b>Hybrid (z-score)</b></td> <td><b>Percent difference</b></td> <td><b>Hybrid (min-max)</b></td> <td><b>Hybrid (z-score)</b></td> <td><b>Percent difference</b></td> <td><b>Hybrid (min-max)</b></td> <td><b>Hybrid (z-score)</b></td> <td><b>Percent difference</b></td> </tr> <tr> <td>SciDocs</td> <td>76.25</td> <td>77.5</td> <td>1.64%</td> <td>99</td> <td>100.5</td> <td>1.52%</td> <td>129.54</td> <td>133.04</td> <td>2.70%</td> </tr> <tr> <td>FIQA</td> <td>80</td> <td>81</td> <td>1.25%</td> <td>104.5</td> <td>105</td> <td>0.48%</td> <td>123.236</td> <td>124</td> <td>0.62%</td> </tr> <tr> <td>NQ</td> <td>117</td> <td>117</td> <td>0%</td> <td>140</td> <td>140</td> <td>0%</td> <td>166.74</td> <td>165.24</td> <td>-0.90%</td> </tr> <tr> <td>ArguAna</td> <td>349</td> <td>349</td> <td>0%</td> <td>382</td> <td>382</td> <td>0%</td> <td>417.975</td> <td>418.475</td> <td>0.12%</td> </tr> <tr> <td>Touche-2020</td> <td>77</td> <td>77.5</td> <td>0.64%</td> <td>100</td> <td>100.5</td> <td>0.50%</td> <td>140</td> <td>140</td> <td>0%</td> </tr> <tr> <td></td> <td></td> <td><b>Average:</b></td> <td>0.70%</td> <td></td> <td><b>Average:</b></td> <td>0.50%</td> <td></td> <td><b>Average:</b></td> <td>0.50%</td> </tr> </table>
```

## Benchmarking z-score normalization for the production environment

Benchmark experiments were conducted using an OpenSearch cluster consisting of a single r6g.8xlarge instance as the coordinator node along with 3 r6g.8xlarge instances as data nodes with 12 shards.

### NDCG@mean

The following table compares NDCG@mean scores for hybrid search using min-max normalization and z-score normalization, showing the percent difference in search relevance across five datasets.

|Dataset	|Hybrid (min-max)	|Hybrid (z-score)	|Percent difference	|
|---	|---	|---	|---	|
|FiQA	|0.3105	|0.3120	|+0.48%	|
|NQ	|0.4563	| 0.4540	|-0.50%|
|ArguAna	|0.4713	|0.4700	|-0.28%	|
|Touche-2020	|0.3788	|0.3719	|-1.82%	|
|	|	|Average	|-0.53%	|

### P@mean

The following table compares precision at mean (P@mean) scores for hybrid search using min-max normalization and z-score normalization, showing the percent difference in search relevance across four datasets.

|Dataset	|Hybrid (min-max)	|Hybrid (z-score)	|Percent difference	|
|---	|---	|---	|---	|
|FiQA	|0.0921	|0.0925	|+0.43%	|
|NQ	|0.0943	| 0.0935	|-0.85%|
|ArguAna	|0.0878	|0.0878	|0%	|

> |Touche2020	|0.2364	|0.2318	|-1.95%	|

Reviewer 5: ```suggestion
|Touche-2020	|0.2364	|0.2318	|-1.95%	|
```

|	|	|Average	|-0.59%	|

### Conclusions

Our production experiments highlight the following aspects of z-score normalization compared to min-max normalization in hybrid search approaches:

**Search quality**:
* Z-score normalization shows a slight decline in search quality, with an average decrease of 0.53% in NDCG@mean scores.
* P@mean results show a similar trend, with a 0.59% average decrease across datasets.
* Only the FIQA dataset demonstrated consistent improvements (+0.48% NDCG@mean, +0.43% P@mean).

> * The largest decline was observed in the Touche2020 dataset (-1.82% NDCG@mean, -1.95% P@mean).

Reviewer 5: ```suggestion
* The largest decline was observed in the Touche-2020 dataset (-1.82% NDCG@mean, -1.95% P@mean).
```

**Trade-offs**:
* Production metrics indicate consistent but minor degradation in search relevance.
* Dataset size appears to influence performance, with larger datasets showing more pronounced decreases.

> * The FIQA dataset demonstrates that z-score can outperform min-max normalization in specific scenarios.

Reviewer 5: ```suggestion
* The FIQA dataset demonstrates that z-score normalization can outperform min-max normalization in specific scenarios.
```

                }
            }
        }
    ]
}
```

Next, create a hybrid query and apply the pipeline to the query:

```json
POST my_index/_search?search_pipeline=z_score-pipeline
{
   "query": {
     "hybrid": [
         {}, // First Query
         {}, // Second Query
         ... // Other Queries
     ] 
   }
}
```

## Benchmarking z-score normalization performance

Benchmark experiments were conducted on an OpenSearch cluster with one **r6g.8xlarge** instance as the coordinator node and three **r6g.8xlarge** instances as data nodes. To evaluate z-score normalization comprehensively, we measured two key metrics across five distinct datasets.  

## Benchmarking z-score normalization for the development environment

Benchmark experiments were conducted using an OpenSearch cluster consisting of a single r6g.8xlarge instance as the coordinator node, along with three r6g.8xlarge instances as data nodes with 1 shard. To assess z -score normalization performance comprehensively, we measured two key metrics across five distinct datasets. For more information about the datasets used, see [Datasets](https://public.ukp.informatik.tu-darmstadt.de/thakur/BEIR/datasets/).

> 

Author: Removed the previous paragraph

Search relevance was quantified using the industry-standard normalized discounted cumulative gain at rank 10 (NDCG@10). We also tracked system performance using search latency measurements. This setup provided a strong foundation for evaluating both search quality and operational efficiency.

### NDCG@10

The following table compares NDCG@10 scores for hybrid search using min-max normalization and z-score normalization, showing the percent difference in search relevance across five datasets.

|Dataset	|Hybrid (min-max)	|Hybrid (z-score)	|Percent difference	|
|---	|---	|---	|---	|
|SciDocs	|0.1591	|0.1633	|+2.45%	|
|FiQA	|0.2747	|0.2768	|+0.77%	|
|NQ	|0.3665	|0.374	|+2.05%	|
|ArguAna	|0.4507	|0.467	|+3.62%	|
|touche2020	|0.841	|0.8542	|+1.54%	|
|	|	|Average	|2.08%	|

### Search latency

The following table presents search latency measurements in milliseconds at different percentiles (p50, p90, and p99) for hybrid search using min-max and z-score normalization. The *Percent difference* columns indicate the relative performance impact of these methods.

> <table> <tr> <th></th> <th colspan="3"><b>p50</b></th> <th colspan="3"><b>p90</b></th> <th colspan="3"><b>p99</b></th> </tr> <tr> <td></td> <td><b>Hybrid (min-max)</b></td> <td><b>Hybrid (z-score)</b></td> <td><b>Percent difference</b></td> <td><b>Hybrid (min-max)</b></td> <td><b>Hybrid (z-score)</b></td> <td><b>Percent difference</b></td> <td><b>Hybrid (min-max)</b></td> <td><b>Hybrid (z-score)</b></td> <td><b>Percent difference</b></td> </tr> <tr> <td>SciDocs</td> <td>76.25</td> <td>77.5</td> <td>1.64%</td> <td>99</td> <td>100.5</td> <td>1.52%</td> <td>129.54</td> <td>133.04</td> <td>2.70%</td> </tr> <tr> <td>FIQA</td> <td>80</td> <td>81</td> <td>1.25%</td> <td>104.5</td> <td>105</td> <td>0.48%</td> <td>123.236</td> <td>124</td> <td>0.62%</td> </tr> <tr> <td>NQ</td> <td>117</td> <td>117</td> <td>0%</td> <td>140</td> <td>140</td> <td>0%</td> <td>166.74</td> <td>165.24</td> <td>-0.90%</td> </tr> <tr> <td>ArguAna</td> <td>349</td> <td>349</td> <td>0%</td> <td>382</td> <td>382</td> <td>0%</td> <td>417.975</td> <td>418.475</td> <td>0.12%</td> </tr> <tr> <td>touche2020</td> <td>77</td> <td>77.5</td> <td>0.64%</td> <td>100</td> <td>100.5</td> <td>0.50%</td> <td>140</td> <td>140</td> <td>0%</td> </tr> <tr> <td></td> <td></td> <td><b>Average:</b></td> <td>0.70%</td> <td></td> <td><b>Average:</b></td> <td>0.50%</td> <td></td> <td><b>Average:</b></td> <td>0.50%</td> </tr> </table>

Reviewer 5: In the table, I changed "touche2020" to "Touche-2020".

- **Latency impact**:

  * Z-score normalization shows a small increase in latency across different percentiles, as shown in the following table.

  |Latency percentile	|Percent difference	|
  |---	|---	|
  |p50	|0.70%	|
  |p90	|0.50%	|
  |p99	|0.50%	|

  * The positive percentages indicate that z-score normalization has slightly higher latency compared to min-max normalization, but the differences are minimal (less than 1% on average).

- **Trade-offs**:

  * There's a slight trade-off between search quality and latency. Z-score normalization offers an improvement in search relevance (2.08% increase in NDCG@10) at the cost of a marginal increase in latency (0.50% to 0.72% across different percentiles).

- **Overall assessment**:

  * Z-score normalization provides a modest improvement in search quality with a negligible impact on latency.
  * The choice between z-score and min-max normalization may depend on your use case. If search relevance is a priority, z-score normalization may be the better option, especially if your system can tolerate a small latency increase.

## Benchmarking z-score normalization for the production environment

> Benchmark experiments were conducted using an OpenSearch cluster consisting of a single r6g.8xlarge instance as the coordinator node, along with three r6g.8xlarge instances as data nodes with 12 shards.

Author: Makes sense. Pushed the change

### NDCG@mean

The following table compares NDCG@mean scores for hybrid search using min-max normalization and z-score normalization, showing the percent difference in search relevance across five datasets.

|Dataset	|Hybrid (min max)	|Hybrid (z score)	|Percent difference	|
|---	|---	|---	|---	|
|FiQA	|0.3105	|0.3120	|+0.48%	|
|NQ	|0.4563	| 0.4540	|-0.50%|
|ArguAna	|0.4713	|0.4700	|-0.28%	|
|touche2020	|0.3788	|0.3719	|-1.82%	|
|	|	|Average	|-0.53%	|

### P@mean

> The following table compares P@mean (Precision at mean) scores for hybrid search using min-max normalization and z-score normalization, showing the percent difference in search relevance across four datasets.

Reviewer: ```suggestion
The following table compares precision at mean (P@mean) scores for hybrid search using min-max normalization and z-score normalization, showing the percent difference in search relevance across four datasets.
```

|Dataset	|Hybrid (min max)	|Hybrid (z score)	|Percent difference	|
|---	|---	|---	|---	|
|FiQA	|0.0921	|0.0925	|+0.43%	|
|NQ	|0.0943	| 0.0935	|-0.85%|
|ArguAna	|0.0878	|0.0878	|0%	|

> |touche2020	|0.2364	|0.2318	|-1.95%	|

Reviewer 5: ```suggestion
|Touche-2020	|0.2364	|0.2318	|-1.95%	|
```

|	|	|Average	|-0.59%	|

### Conclusions

Our production experiments highlight the following aspects of Z-score normalization compared to min-max normalization in hybrid search approaches:

**Search quality**:
* Z-score normalization shows a slight decline in search quality, with an average decrease of 0.53% in NDCG@mean scores
* P@mean results show a similar trend with a 0.59% average decrease across datasets
* Only FIQA dataset demonstrated consistent improvements (+0.48% NDCG@mean, +0.43% P@mean)
* Largest decline observed in Touche2020 dataset (-1.82% NDCG@mean, -1.95% P@mean)

**Trade-offs**:

> * Production metrics indicate consistent but minor degradation in search relevance

Reviewer 5: ```suggestion
* Production metrics indicate consistent but minor degradation in search relevance.
```

> * Dataset size appears to influence performance, with larger datasets showing more pronounced decreases  

Reviewer 5: ```suggestion
* Dataset size appears to influence performance, with larger datasets showing more pronounced decreases.
```

> * FIQA dataset demonstrates that Z-score can outperform min-max in specific scenarios

Reviewer 5: ```suggestion
* The FIQA dataset demonstrates that z-score normalization can outperform min-max normalization in specific scenarios.
```

> * Performance impact varies significantly across different datasets

Reviewer 5: ```suggestion
* Performance impact varies significantly across different datasets.
```

**Overall assessment**:

> * Z-score normalization performs marginally below min-max normalization in production settings

Reviewer 5: ```suggestion
* Z-score normalization performs marginally below min-max normalization in production settings.
```

> * Results suggest dataset characteristics significantly influence normalization effectiveness

Reviewer 5: ```suggestion
* Results suggest dataset characteristics significantly influence normalization effectiveness.
```

> * Implementation decisions should prioritize specific use case requirements and dataset characteristics

Reviewer 5: ```suggestion
* Implementation decisions should prioritize specific use case requirements and dataset characteristics.
```

> These findings indicate that while Z-score normalization remains a viable option, min-max normalization demonstrates more reliable performance in production scenarios. The choice between methods should be based on thorough testing with specific datasets and use cases.

Reviewer 5: ```suggestion
These findings indicate that while z-score normalization remains a viable option, min-max normalization demonstrates more reliable performance in production scenarios. The choice between methods should be based on thorough testing with specific datasets and use cases.
```
